contracts/base/BaseLiquidityManagement.sol

```diff
@@ -96,39 +99,77 @@ contract BaseLiquidityManagement is IBaseLiquidityManagement, SafeCallback {
-        // if callerDelta < 0, existing fees were re-invested AND net new tokens are required for the liquidity increase
-        // if callerDelta == 0, existing fees were reinvested (autocompounded)
-        // if callerDelta > 0, some but not all existing fees were used to increase liquidity. Any remainder is added to the position's owed tokens
-        if (callerDelta.amount0() > 0) {
-            position.tokensOwed0 += uint128(callerDelta.amount0());
-            range.key.currency0.take(manager, address(this), uint128(callerDelta.amount0()), true);
-            callerDelta = toBalanceDelta(0, callerDelta.amount1());
-        } else {
-            position.tokensOwed0 = 0;
-        }
+        // Calculate the portion of the liquidityDelta that is attributable to the caller.
+        // We must account for fees that might be owed to other users on the same range.
+        (uint256 feeGrowthInside0X128, uint256 feeGrowthInside1X128) =
+            manager.getFeeGrowthInside(range.poolKey.toId(), range.tickLower, range.tickUpper);
 
-        if (callerDelta.amount1() > 0) {
-            position.tokensOwed1 += uint128(callerDelta.amount1());
-            range.key.currency1.take(manager, address(this), uint128(callerDelta.amount1()), true);
-            callerDelta = toBalanceDelta(callerDelta.amount0(), 0);
-        } else {
-            position.tokensOwed1 = 0;
+        BalanceDelta callerFeesAccrued = FeeMath.getFeesOwed(
+            feeGrowthInside0X128,
+            feeGrowthInside1X128,
+            position.feeGrowthInside0LastX128,
+            position.feeGrowthInside1LastX128,
+            position.liquidity
+        );
+
+        // Calculate the accurate callerDelta.
+        // If the totalFeesAccrued equals the callerFeesAccrued then the callerDelta is just the liquidityDelta.
+        // If the totalFeesAccrued after the add liquidity action is greater than only the caller's earned fees, we must account for the difference.
+        (int128 callerDelta0, int128 callerDelta1) =
+            _accountNewCallerDeltas(liquidityDelta, totalFeesAccrued, callerFeesAccrued, range.poolKey);
+
+        // Update position storage (tokensOwed from fees, liquidity added, and new feeGrowthInside).
+        // if callerDelta <= 0, then tokensOwed0 and tokensOwed1 should be zero'd out as all fees were re-invested into a new position.
+        // if callerDelta > 0, then even after re-investing old fees, the caller still has some fees to collect that were not added into the position so they are accounted.
+        uint128 tokensOwed0 = callerDelta0 > 0 ? uint128(callerDelta0) : 0;
+        uint128 tokensOwed1 = callerDelta1 > 0 ? uint128(callerDelta1) : 0;
+
+        position.updateTokensOwed(tokensOwed0, tokensOwed1);
+        position.add(liquidityToAdd);
+        position.updateFeeGrowthInside(feeGrowthInside0X128, feeGrowthInside1X128);
+
+        return toBalanceDelta(callerDelta0, callerDelta1);
```

Comment:
is this set to zero when `callerDelta > 0` (and we credit the remainder fees to `position.tokensOwed`?)